webapp/sources/rudder/rudder-core/src/main/scala/com/normation/rudder/ncf/TechniqueCompiler.scala

```diff
@@ -0,0 +1,1197 @@
+
+  // an error that should be displayed to user
+  final case class UserError(code: Int, msg: String, stdout: String, stderr: String) extends RuddercError
+  // a rudderc error that is not recoverable by the user and should lead to a fallback
+  final case class Fail(code: Int, msg: String, stdout: String, stderr: String)      extends RuddercError
+
+  // biased it toward system error for now, it can change when rudderc is mature enough
+  def parseToError(code: Int, stdout: String, stderr: String): RuddercResult = {
+    // todo: parse a rudderc output and get an error from it
+    if (code == 0) {
+      Ok("", stdout, stderr)
+    } else {
+      Fail(code, "", stdout, stderr)
+    }
+  }
+}
+
+/*
+ * Option for rudder, like verbosity, etc
+ */
+final case class RuddercOptions(
+    generatePS1: Boolean,
+    verbose:     Boolean
+)
+
+sealed trait NcfError extends RudderError {
+  def message:   String
+  def exception: Option[Throwable]
+  def msg = message
+}
+
+final case class IOError(message: String, exception: Option[Throwable])              extends NcfError
+final case class TechniqueUpdateError(message: String, exception: Option[Throwable]) extends NcfError
+final case class MethodNotFound(message: String, exception: Option[Throwable])       extends NcfError
+
+/*
+ * A trait that call Rudderc and translate it's output to things understandable by Rudder.
+ * Can be mocked for tests
+ */
+trait RuddercService {
+
+  // techniqueDir: absolute path to technique main directory (techniqueId/version)
+  def compile(techniqueDir: File, options: RuddercOptions): IOResult[RuddercResult]
+}
+
+class RuddercServiceImpl(
+    val ruddercCmd:      String,
+    ruddercFallbackCode: Int,
+    killTimeout:         Duration
+) extends RuddercService {
+
+  override def compile(techniqueDir: File, options: RuddercOptions): IOResult[RuddercResult] = {
+    val cmd = buildCmdLine(techniqueDir, options)
+
+    for {
+      _       <- RuddercLogger.debug(s"Run rudderc: ${cmd.display}")
+      time_0  <- currentTimeNanos
+      p       <- RunNuCommand.run(cmd) // this can't fail, errors are captured in return code
+      r       <- p.await.timeout(killTimeout).flatMap {
+                   case Some(ok) =>
+                     ok.succeed
+                   case None     =>
+                     val error = s"Rudderc ${cmd.display} timed out after ${killTimeout.render}"
+                     RuddercLogger.error(error) *> CmdResult(ruddercFallbackCode, "", error).succeed
+                 }
+      c        = translateReturnCode(techniqueDir.pathAsString, r)
+      _       <- logReturnCode(c)
+      time_1  <- currentTimeNanos
+      duration = time_1 - time_0
+      _       <- RuddercLogger.debug(
+                   s"Done in ${duration / 1000} us: ${cmd.display}"
+                 )
+    } yield {
+      c
+    }
+  }
+
+  def buildCmdLine(techniquePath: File, options: RuddercOptions): Cmd = {
+    // todo: real param list here
```

Comment:
indeed!

Reply:
For verbosity rudderc supports:

* `-q`: only show warnings and errors
* no verbosity parameter : show error+warn+info
* `-v` : show error+warn+info+debug
* `-vv` : show error+warn+info+debug+trace

Reply:
The current way to build the techniques is to run `rudderc build` from the technique directory. There is no way to select output backends.

This could be modified though.